### PR TITLE
Add Skew Normal distribution

### PR DESCRIPTION
**Pull request by leandrolcampos**
This PR provides an implementation for the Skew Normal distribution proposed by Fernández and Steel (1998)[1]. It addresses the issue [#399](https://github.com/tensorflow/probability/issues/399).

The Skew Normal generalizes the Normal distribution with an additional shape parameter. It is parameterized by location `loc`, scale `scale`, and shape `skewness`. If the skewness is above one, the distribution becomes positively skewed (or right-skewed). If the skewness is greater than zero and less than one, the distribution becomes negatively skewed (or left-skewed). A skewness equal to one results in a Normal distribution.

I tried to follow all the style decisions provided in the [style guide](https://github.com/tensorflow/probability/blob/main/STYLE_GUIDE.md). But in some cases I couldn't follow the TensorFlow Probability-specific conventions (8), (16) or (24) because one or more tests failed.

This PR, including the test suite, was inspired by the available implementation of the Generalized Normal distribution.

Any feedback is appreciated.

[1] Carmen Fernández and Mark F. J. Steel. On Bayesian modeling of fat tails and skewness. _Journal of the American Statistical Association_, 93(441), 359-371, 1998.

### CODE REVIEW

File: tensorflow_probability/python/distributions/skew_normal.py [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+  abs_skewness = tf.math.abs(skewness)
+  adj_scale = tf.math.abs(scale) * tf.where(
+      cond, x=-tf.math.reciprocal(abs_skewness), y=abs_skewness)
+
+  return loc + adj_scale * tf.cast(
+      tf.math.sqrt(two * gamma_quantile), dtype=loc.dtype)
+
+
+class SkewNormal(distribution.AutoCompositeTensorDistribution):
+  """The Skew Normal distribution.
```
Comment (SiegeLordEx) [inline tensorflow_probability/python/distributions/skew_normal.py:122]: I'm a little concerned that we're calling this 'The' skew normal distribution, when there's at least two common variants of this. This one appears to be by the 'type 2' from gamlss, while when people search for skew normal, e.g., on wikipedia they get the 'type 1' (https://en.wikipedia.org/wiki/Skew_normal_distribution). Wikipedia also has https://en.wikipedia.org/wiki/Split_normal_distribution which looks similar to this formulation, but not quite the same. I note you cite `Fernández-Steel distribution` below (I couldn't get a copy of that manuscript, so I don't know the details).

Do you have any insight about what best to call it?

File: tensorflow_probability/python/distributions/skew_normal.py [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+  # X ~ Normal(loc=0, scale=1) => 2 * X**2 ~ Gamma(alpha=0.5, beta=1)
+  probs = (one - value * (one + squared_skewness)) * tf.where(
+      cond, x=one, y=-tf.math.reciprocal(squared_skewness))
+  gamma_quantile = tfp_math.igammainv(half, p=probs)
+
+  abs_skewness = tf.math.abs(skewness)
+  adj_scale = tf.math.abs(scale) * tf.where(
+      cond, x=-tf.math.reciprocal(abs_skewness), y=abs_skewness)
+
+  return loc + adj_scale * tf.cast(
```
[the remark this answers is not in the document]
Comment (leandrolcampos) [reply tensorflow_probability/python/distributions/skew_normal.py:117]: When I try to remove this cast, the test fails with the following message:

> (...)
INFO: Analyzed 3 targets (0 packages loaded, 0 targets configured).
INFO: Found 3 test targets...
FAIL: //tensorflow_probability/python/distributions:skew_normal_test.numpy (see /home/leandro/.cache/bazel/_bazel_leandro/b8274678e6a17bc831bbd9e0d37a2c28/execroot/tensorflow_probability/bazel-out/k8-opt/testlogs/tensorflow_probability/python/distributions/skew_normal_test.numpy/test.log)
(...)
FAIL: testSampleLikeArgsGetDistDType (__main__.SkewNormalTestDynamicShapeFloat32)
SkewNormalTestDynamicShapeFloat32.testSampleLikeArgsGetDistDType
Traceback (most recent call last):
  File "/home/leandro/.cache/bazel/_bazel_leandro/b8274678e6a17bc831bbd9e0d37a2c28/sandbox/linux-sandbox/874/execroot/tensorflow_probability/bazel-out/k8-opt/bin/tensorflow_probability/python/distributions/skew_normal_test.numpy.runfiles/tensorflow_probability/tensorflow_probability/python/distributions/_numpy/_generated_skew_normal_test.py", line 110, in testSampleLikeArgsGetDistDType
    self.assertEqual(self.dtype, getattr(dist, method)(1).dtype)
  File "/home/leandro/.cache/bazel/_bazel_leandro/b8274678e6a17bc831bbd9e0d37a2c28/sandbox/linux-sandbox/874/execroot/tensorflow_probability/bazel-out/k8-opt/bin/tensorflow_probability/python/distributions/skew_normal_test.numpy.runfiles/tensorflow_probability/tensorflow_probability/python/internal/backend/numpy/test_lib.py", line 85, in assertEqual
    return super(TestCase, self).assertEqual(first, second, msg)
AssertionError: <class 'numpy.float32'> != dtype('float64')
FAIL: testSampleLikeArgsGetDistDType (__main__.SkewNormalTestStaticShapeFloat32)
SkewNormalTestStaticShapeFloat32.testSampleLikeArgsGetDistDType
Traceback (most recent call last):
  File "/home/leandro/.cache/bazel/_bazel_leandro/b8274678e6a17bc831bbd9e0d37a2c28/sandbox/linux-sandbox/874/execroot/tensorflow_probability/bazel-out/k8-opt/bin/tensorflow_probability/python/distributions/skew_normal_test.numpy.runfiles/tensorflow_probability/tensorflow_probability/python/distributions/_numpy/_generated_skew_normal_test.py", line 110, in testSampleLikeArgsGetDistDType
    self.assertEqual(self.dtype, getattr(dist, method)(1).dtype)
  File "/home/leandro/.cache/bazel/_bazel_leandro/b8274678e6a17bc831bbd9e0d37a2c28/sandbox/linux-sandbox/874/execroot/tensorflow_probability/bazel-out/k8-opt/bin/tensorflow_probability/python/distributions/skew_normal_test.numpy.runfiles/tensorflow_probability/tensorflow_probability/python/internal/backend/numpy/test_lib.py", line 85, in assertEqual
    return super(TestCase, self).assertEqual(first, second, msg)
AssertionError: <class 'numpy.float32'> != dtype('float64')

File: tensorflow_probability/python/distributions/skew_normal.py [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+
+    half = tf.constant(0.5, dtype=self.dtype)
+    two = tf.constant(2., dtype=self.dtype)
+    pi = tf.constant(np.pi, dtype=self.dtype)
+
+    z = standardize(value, loc=loc, scale=scale, skewness=skewness)
+
+    log_unnormalized = -half * tf.math.square(z)
+    log_normalization = (
+        tf.cast(log1psquare(skewness), dtype=self.dtype) -
```
[the remark this answers is not in the document]
Comment (leandrolcampos) [reply tensorflow_probability/python/distributions/skew_normal.py:366]: When I try to remove this tf.cast, the test also fails with the same message:

> (...)
INFO: Analyzed 3 targets (0 packages loaded, 0 targets configured).
INFO: Found 3 test targets...
FAIL: //tensorflow_probability/python/distributions:skew_normal_test.numpy (see /home/leandro/.cache/bazel/_bazel_leandro/b8274678e6a17bc831bbd9e0d37a2c28/execroot/tensorflow_probability/bazel-out/k8-opt/testlogs/tensorflow_probability/python/distributions/skew_normal_test.numpy/test.log)
(...)
FAIL: testSampleLikeArgsGetDistDType (__main__.SkewNormalTestDynamicShapeFloat32)
SkewNormalTestDynamicShapeFloat32.testSampleLikeArgsGetDistDType
Traceback (most recent call last):
  File "/home/leandro/.cache/bazel/_bazel_leandro/b8274678e6a17bc831bbd9e0d37a2c28/sandbox/linux-sandbox/883/execroot/tensorflow_probability/bazel-out/k8-opt/bin/tensorflow_probability/python/distributions/skew_normal_test.numpy.runfiles/tensorflow_probability/tensorflow_probability/python/distributions/_numpy/_generated_skew_normal_test.py", line 110, in testSampleLikeArgsGetDistDType
    self.assertEqual(self.dtype, getattr(dist, method)(1).dtype)
  File "/home/leandro/.cache/bazel/_bazel_leandro/b8274678e6a17bc831bbd9e0d37a2c28/sandbox/linux-sandbox/883/execroot/tensorflow_probability/bazel-out/k8-opt/bin/tensorflow_probability/python/distributions/skew_normal_test.numpy.runfiles/tensorflow_probability/tensorflow_probability/python/internal/backend/numpy/test_lib.py", line 85, in assertEqual
    return super(TestCase, self).assertEqual(first, second, msg)
AssertionError: <class 'numpy.float32'> != dtype('float64')
FAIL: testSampleLikeArgsGetDistDType (__main__.SkewNormalTestStaticShapeFloat32)
SkewNormalTestStaticShapeFloat32.testSampleLikeArgsGetDistDType
Traceback (most recent call last):
  File "/home/leandro/.cache/bazel/_bazel_leandro/b8274678e6a17bc831bbd9e0d37a2c28/sandbox/linux-sandbox/883/execroot/tensorflow_probability/bazel-out/k8-opt/bin/tensorflow_probability/python/distributions/skew_normal_test.numpy.runfiles/tensorflow_probability/tensorflow_probability/python/distributions/_numpy/_generated_skew_normal_test.py", line 110, in testSampleLikeArgsGetDistDType
    self.assertEqual(self.dtype, getattr(dist, method)(1).dtype)
  File "/home/leandro/.cache/bazel/_bazel_leandro/b8274678e6a17bc831bbd9e0d37a2c28/sandbox/linux-sandbox/883/execroot/tensorflow_probability/bazel-out/k8-opt/bin/tensorflow_probability/python/distributions/skew_normal_test.numpy.runfiles/tensorflow_probability/tensorflow_probability/python/internal/backend/numpy/test_lib.py", line 85, in assertEqual
    return super(TestCase, self).assertEqual(first, second, msg)
AssertionError: <class 'numpy.float32'> != dtype('float64')

File: tensorflow_probability/python/distributions/skew_normal.py (hunk 1 above)
Comment (leandrolcampos) [reply tensorflow_probability/python/distributions/skew_normal.py:122]: Thanks for your feedback!

We can call this “The Two-Piece Normal distribution” [(Robert A. Rigby et al., 2019, Section 18.3.5, p380)][3].

In fact, according to [Rubio and Steel (2020, Section “Who discovered them?”)][4], the two-piece normal distribution “was introduced in a general framework by Fernández and Steel under an alternative parameterisation”. They refer to the formulation proposed by [Fernández and Steel (1998)][2]. So, to be more specific, we can call this distribution “The Two-Piece Normal distribution under the parameterization proposed by [Fernández and Steel (1998)][2]”. And we can name the class `TwoPieceNormal`. What do you think? If you agree, I can make the necessary changes.

We can see the mathematical equivalence between the parameterization proposed by [Fernández and Steel (1998)][2] and the main parameterization discussed in https://en.wikipedia.org/wiki/Split_normal_distribution. Consider that the two-piece normal distribution is built joining at the mode the corresponding halves of two normal distributions with the same mode but different standard deviation. Under the former parameterization, the left (or right) hand side standard deviation is given by `sigma / skewness` (or `sigma * skewness`), where `sigma` and `skewness` are, respectively, the scale and shape parameters of the two-piece normal distribution. Under the latter parameterization, the left (or right) hand side standard deviation is given directly by `sigma_1` (or `sigma_2`). If we use the relationships `sigma_1 = sigma / skewness` and `sigma_2 = sigma * skewness`, then we can go from one parameterization to the other.

The other option consists of maintaining the class name `SkewNormal` but presenting the distribution as "The Fernández-Steel Skew Normal distribution", following [Castillo et al. (2011)][1]. In fact, there is no official version of the skew normal distribution. And the class name `SkewNormal` is better at informing the purpose and main features of this distribution than the class name `TwoPieceNormal` (although the latter is more specific).

[1] Nabor O. Castillo et al. On the Fernández-Steel distribution: Inference and application. _Computational Statistics & Data Analysis_, 55(11), 2951-2961, 2011.

[2] Carmen Fernández and Mark F. J. Steel. On Bayesian modeling of fat tails and skewness. _Journal of the American Statistical Association_, 93(441), 359-371, 1998.

[3] Robert A. Rigby et al. _Distributions for modeling location, scale, and shape: Using GAMLSS in R_. Chapman and Hall/CRC, 2019.

[4] Francisco J. Rubio and Mark F. J. Steel. The family of two‐piece distributions. _Significance_, 17(1), 12-13, 2020.
https://rss.onlinelibrary.wiley.com/doi/full/10.1111/j.1740-9713.2020.01352.x

Comment (SiegeLordEx) [reply tensorflow_probability/python/distributions/skew_normal.py:122]: Thanks for looking into it. Let's do this:

1. Rename it to TwoPieceNormal.
2. Add the sentence you proposed (The Two-Piece Normal distribution under the parameterization proposed by [Fernández and Steel (1998)][2])
3. Somewhere in the docstring we should have a section saying that this distribution has also been called the SkewNormal and SplitNormal and cite wikipedia and GAMLSS.

The docstring change should mitigate the somewhat indirect name, so if people seach for skewnormal they might still be able to find this distribution.

Comment (leandrolcampos) [reply tensorflow_probability/python/distributions/skew_normal.py]: I made the necessary changes as we agreed.

**Review by SiegeLordEx — APPROVED** [review]

**Comment by SiegeLordEx** [comment]
Thanks again. I'll send this along to our internal review system and once that's done, we'll merge this.

**Comment by SiegeLordEx** [comment]
In internal review we observed that the samples of this distributions are not actually reparameterizable, due to the discrete bernoulli used to pick which piece to sample from. We removed/changed the code that implied otherwise. If you're interested in fixing that, it's still possible to implement reparameterizable gradients for this via implicit methods. A good example of that is the https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/distributions/truncated_normal.py code. The overall approach is described here: https://arxiv.org/abs/1805.08498

**Comment by mihailyanchev** [comment]
@leandrolcampos Thank you for your contribution. I had a quick question. I noticed that the textbook rule on the relative positions of mean and median does not apply here - for right/positive skewness the mean is greater than the median and vice versa. Is this expected given this specific implementation? I know this textbook rule is not reliable (https://www.tandfonline.com/doi/pdf/10.1080/10691898.2005.11910556), but just wondered if you have looked into this by any chance?

`test_dist = SkewNormal(loc=0, scale=1, skewness=1.25)
print(np.array(test_dist.quantile(0.5)))
0.28443122
`

**Comment by leandrolcampos** [comment]
Hi @mihailyanchev! I hope this message finds you well. And thanks for your feedback!

In the case of Two-Piece Normal distribution, the location parameter is different from the mean. In fact, it defines the mode. Please consider the following code block:

```python
import tensorflow_probability as tfp
tfd = tfp.distributions

# If skewness is greater than zero and less than one, dist becomes left-skewed 
# (or negatively skewed).
dist = tfd.TwoPieceNormal(loc=0., scale=1., skewness=0.75)

mode = dist.mode() # == loc
median = dist.quantile(0.5)
mean = dist.mean()

print(mean < median < mode) # True
print(mean, median, mode) # -0.46543267, -0.3702539, 0.0

# If skewness is above one, the distribution becomes right-skewed
# (or positively skewed).
dist = tfd.TwoPieceNormal(loc=0., scale=1., skewness=1.33)

mode = dist.mode() # == loc
median = dist.quantile(0.5)
mean = dist.mean()

print(mode < median < mean) # True
print(mode, median, mean) # 0.0, 0.3668836, 0.46127328
```

Another reference: https://en.wikipedia.org/wiki/Skewness

> @leandrolcampos Thank you for your contribution. I had a quick question. I noticed that the textbook rule on the relative positions of mean and median does not apply here - for right/positive skewness the mean is greater than the median and vice versa. Is this expected given this specific implementation? I know this textbook rule is not reliable (https://www.tandfonline.com/doi/pdf/10.1080/10691898.2005.11910556), but just wondered if you have looked into this by any chance?
> 
> `test_dist = SkewNormal(loc=0, scale=1, skewness=1.25)` `print(np.array(test_dist.quantile(0.5)))` `0.28443122`

**Comment by mihailyanchev** [comment]
@leandrolcampos thanks for the quick reply. This is very helpful. I did not realize loc defines the mode and this makes it clear.

**Comment by leandrolcampos** [comment]
> In internal review we observed that the samples of this distributions are not actually reparameterizable, due to the discrete bernoulli used to pick which piece to sample from. We removed/changed the code that implied otherwise. If you're interested in fixing that, it's still possible to implement reparameterizable gradients for this via implicit methods. A good example of that is the https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/distributions/truncated_normal.py code. The overall approach is described here: https://arxiv.org/abs/1805.08498

Hi @SiegeLordEx, I tried to make the Two-Piece Normal distribution differentiable w.r.t. all parameters in this PR: https://github.com/tensorflow/probability/pull/1541

**Pull request merged**